### PR TITLE
feat: Implemented CopyGraph

### PR DESCRIPTION
**Pull request by sammychinedu2ky**
Fixes https://github.com/oras-project/oras-dotnet/issues/5

### CODE REVIEW

File: Oras/Oras.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            var root = await src.ResolveAsync(srcRef, cancellationToken);
+            await CopyGraphAsync(src, dst, root, cancellationToken);
+            await dst.TagAsync(root, dstRef, cancellationToken);
+            return root;
+        }
+
+        private async Task CopyGraphAsync(ITarget src, ITarget dst, Descriptor node, CancellationToken cancellationToken)
+        {
+            // check if node exists in target
+            if (await dst.ExistsAsync(node, cancellationToken))
```
Comment (Wwwsylvia) [inline Oras/Oras.cs:46]: Why do we need to copy when the content exists in dst?

File: Oras/Oras.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                // retrieve child nodes
+                var childNodes = await StorageUtility.SuccessorsAsync(src, node, cancellationToken);
+                // obtain data stream
+                var dataStream = new MemoryStream(await StorageUtility.FetchAllAsync(src, node, cancellationToken));
```
Comment (Wwwsylvia) [inline Oras/Oras.cs:51]: Can we just use `src.FetchAsync`?

File: Oras/Oras.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            return root;
+        }
+
+        private async Task CopyGraphAsync(ITarget src, ITarget dst, Descriptor node, CancellationToken cancellationToken)
+        {
+            // check if node exists in target
+            if (await dst.ExistsAsync(node, cancellationToken))
+            {
+                // retrieve child nodes
+                var childNodes = await StorageUtility.SuccessorsAsync(src, node, cancellationToken);
```
Comment (Wwwsylvia) [inline Oras/Oras.cs:49]: Note that `SuccessorAsync` internally calls `FetchAsync`. Which means a node will actually be fetched multiple times in the current implementation.

File: Oras/Oras.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                dstRef = srcRef;
+            }
+            var root = await src.ResolveAsync(srcRef, cancellationToken);
+            await CopyGraphAsync(src, dst, root, cancellationToken);
+            await dst.TagAsync(root, dstRef, cancellationToken);
+            return root;
+        }
+
+        private async Task CopyGraphAsync(ITarget src, ITarget dst, Descriptor node, CancellationToken cancellationToken)
```
Comment (Wwwsylvia) [inline Oras/Oras.cs:43]: This function should be public.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: Oras/Copy.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+namespace Oras
+{
+    public class Copy
+    {
+        // Copy copies a rooted directed acyclic graph (DAG) with the tagged root node
+        // in the source Target to the destination Target.
+        // The destination reference will be the same as the source reference if the
+        // destination reference is left blank.
+        // Returns the descriptor of the root node on successful copy.
+        public async Task<Descriptor> CopyAsync(ITarget src, string srcRef, ITarget dst, string dstRef, CancellationToken cancellationToken)
```
Comment (Wwwsylvia) [inline Oras/Copy.cs:17]: Should this be a static method?

File: Oras/Copy.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace Oras
+{
+    public class Copy
+    {
+        // Copy copies a rooted directed acyclic graph (DAG) with the tagged root node
+        // in the source Target to the destination Target.
+        // The destination reference will be the same as the source reference if the
+        // destination reference is left blank.
+        // Returns the descriptor of the root node on successful copy.
```
Comment (Wwwsylvia) [inline Oras/Copy.cs:16]: These are not C# style comment docs.

File: Oras/Copy.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            return root;
+        }
+
+        public async Task CopyGraphAsync(ITarget src, ITarget dst, Descriptor node, CancellationToken cancellationToken)
+        {
+            // check if node exists in target
+            if (!await dst.ExistsAsync(node, cancellationToken))
+            {
+                // retrieve child nodes
+                var childNodes = await StorageUtility.SuccessorsAsync(src, node, cancellationToken);
```
Comment (Wwwsylvia) [inline Oras/Copy.cs:43]: Why not just call it `successors`?

File: Oras/Copy.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                        await CopyGraphAsync(src, dst, childNode, cancellationToken);
+                    }
+                    await dst.PushAsync(node, dataStream, cancellationToken);
+                }
+                else
+                {
+                    await dst.PushAsync(node, dataStream, cancellationToken);
+                }
+            }
+        }
```
Comment (Wwwsylvia) [inline Oras/Copy.cs:60]: Can we rewrite this method in an iterative (non-recursive) style?

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: Oras/Copy.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (Wwwsylvia) [inline Oras/Copy.cs:70]: How about 
```suggestion
                if (successors != null)
                {
                    foreach (var childNode in successors)
                    {
                        await CopyGraphAsync(src, dst, childNode, cancellationToken);
                    }
                }
                await dst.PushAsync(node, dataStream, cancellationToken);
            }
        }
```

Comment (sammychinedu2ky) [reply Oras/Copy.cs:70]: yh that's right

**Pull request closed without merging**